boundlessgeo/exchange
# Autocomplete search

## JIRA Ticket
GVSD-10064

## Description
The old way of doing autocompletes was outdated and no longer working. This PR introduces autocompletion using Elasticsearch and an updated select2 v4.0+ library.

Limitations: The styling isn't perfect or exactly the same as it used to be. As such the main search and header search may look a bit odd due to select2's dropdown as well as missing the font awesome search icon. Additionally, the smooth transition for the header bar search does not work with select2.

NOTE: This PR must be merged with the PR to GeoNode (https://github.com/boundlessgeo/geonode/pull/274) and PR to geonode-elasticsearch (https://github.com/boundlessgeo/geonode-elasticsearch/pull/81).

## TODO

## Steps to Test or Reproduce
The following elements within Exchange should have autocomplete functioning properly, both to make selections and submit forms (follow through the entire way to ensure things are working after selection).

1. Main page search - should be a single select which redirects you to the search page for the selection you make
2. Header search - should be a single select which redirects you to the search page for the selection you make
3. Permissions - should be a multiple select which allows you to pick the users and groups you want to give certain permissions to for the corresponding data (make sure permissions are in fact applied)

## Setup Environment with PR

1. Cleanup previous state

```bash
make purge
```

2. Checkout PR

__Using git command (command line interface)__
```bash
pr_id=ADD_PR_NUMBER_HERE
git checkout master
git branch -D $pr_id
git fetch
git fetch origin pull/$pr_id/head:$pr_id
git checkout $pr_id
git submodule init
git submodule update --remote --recursive
```

__Using [Gitkraken](https://www.gitkraken.com/)__

+ In GitKraken, right click on the pull request you want to review. Select Add Remote and Checkout (the PR).

3. Start exchange

```bash
make start
```

4. Exchange Healthcheck

```bash
docker inspect --format '{{ .State.Health.Status }}' exchange
```

__NOTE:__ Only continue the following steps if the output from the above command is `healthy`. You may have to wait 
a few minutes.

---
@boundlessgeo/bex-qa
